---
title: "Setbp1_PathwayAnalysisCortexKidney_07"
date: '2022-09-21'
output:     
  html_document:
      toc: true
      toc_depth: 2
      toc_float: true
editor_options: 
  chunk_output_type: inline
---

This analysis comes after 'Setbp1_CellTypesDGEcortex_06.Rmd', 'Setbp1_CellTypesDGEkidney_06.Rmd'.

## Code: 

#### Loading in libraries:
```{r}
set.seed(2178)
library(Seurat)
library(gprofiler2)
library(tidyverse)
library(lattice)
library(msigdbr)
library(ggplot2)
library(rrvgo)
library(dplyr)
library(readr)
library(here)
library(lintr)
library(styler)
source(here("src/functions/functions.R"))
```

#### loading outputs at log2FC of 0.1 from previous script:
```{r}
# creating vector of DEG file paths for for loop
files_list <- list.files(here("results/DEG"),
  pattern = "het_DEG01",
  full.names = TRUE
)
files_list

# loading in all DEG files:
for (i in files_list) {
  load(i)
}
```

#### formatting all DEG into a single data frame:
Combining the p-value, gene, and differential expression data, while adding a cell type tag and binding all cell-type specific data frames together
```{r}
cortex_list <- list(
  inhibitory_neurons = inhibitory_neurons_het,
  microglia = microglia_het,
  excitatory_neurons = excitatory_neurons_het,
  opcs = opcs_het,
  oligodendrocytes = oligodendrocytes_het,
  astrocytes = astrocytes_het,
  mural_cells = mural_cells_het,
  fibroblasts = fibroblasts_het
)

# wrangling data
colval <- c("p_val", "gene", "diffexpressed")
cortex_list <- lapply(cortex_list, function(x) x[, intersect(names(x), colval), drop = FALSE])

## grab list names
n <- names(cortex_list)

## add new columns for "celltype" with the name of the list as the value
cortex_list <- lapply(names(cortex_list), function(x) {
  cortex_list[[x]] %>%
    mutate(celltype = x)
})

## combining list into single dataframe
het_cortex_DEGS <- bind_rows(cortex_list)
rownames(het_cortex_DEGS) <- NULL

# saving dataframe as a csv to outputs folder
write.csv(het_cortex_DEGS, file = here("results/seurat/Pathways_07/het_cortex_DEG.csv"))
```

```{r}
# creating vector of DEG file paths for for loop
files_list <- list.files(here("results/DEG"),
  pattern = "het_kid_DEG01", # grabbing kidney fibroblasts
  full.names = TRUE
)
files_list

# loading in all DEG files:
for (i in files_list) {
  load(i)
}

kidney_list <- list(
  bcell = bcell_het,
  CDPC = CDPC_het,
  CDICA = CDIC_typeA_het,
  CDICB = CDIC_typeB_het,
  DCT = DCT_het,
  DLH = DLH_het,
  endothelial = endo_het,
  fibroblasts = fibro_het,
  pericytes = pericytes_het,
  LOH = LOH_het,
  macrophages = macro_het,
  PCTS1 = PCTS1_het,
  PCTS2 = PCTS2_het,
  podocytes = podo_het,
  PST = PST_het,
  smooth_muscle_cells = smcs_het,
  proximal_tubule = prox_tubule_het
)


# wrangling data
colval <- c("p_val", "gene", "diffexpressed")
kidney_list <- lapply(kidney_list, function(x) x[, intersect(names(x), colval), drop = FALSE])

## grab list names
n <- names(kidney_list)

## add new columns for "celltype" with the name of the list as the value
kidney_list <- lapply(names(kidney_list), function(x) {
  kidney_list[[x]] %>%
    mutate(celltype = x)
})

## combining list into single dataframe
het_kidney_DEGS <- bind_rows(kidney_list)
rownames(het_kidney_DEGS) <- NULL

# saving dataframe as a csv to outputs folder
write.csv(het_kidney_DEGS, file = here("results/seurat/Pathways_07/het_kidney_DEG.csv"))
```

#### Pathway Analysis Function (gprofiler2)
*credit to Tabea and Lizzy for some function help, assitance with pathway analysis and code review, also received help from lizzy finding gene sets in msigdbr for me and going through GSEA vs FEA and the analytical difference to consider* 

Arguments needed to be passed to the function
* organism; can be "hsapiens" or "mmusculus", must be specified
* user_threshold; default is 0.05 if not specified 
* correction_method; default is bonferroni if not specified

Note that the default pathway size is 10 < term_size < 1000 and the filtered fea result 'fea_result_filt', this gives pathways that are not too large or too small to avoid arbitrary annotations. The top 50 pathways are returned by default in the function, if this is desired to be changed needs to be changed in the function itself.

The Data source for functional enrichment analysis was set to NULL, which means the default sources will be incorporated. As of October 2022 those included: 
* Gene onotlogy sources - GO molecular function, GO cellular component, and GO biological process
* Biological pathway sources - KEGG, Reactome, and WikiPathways
* regulatory motifs in DNA - TRANSFAC and miRTarBase
* Protein Databases - CORUM (Human Protein Atlas excluded for mouse analysis)
* Human Phenotype ontology - HP (still included even in the case of a mouse analysis)

More specific information on data version of the above sources is below: 

GO:MF  annotations: BioMart
classes: releases/2022-03-22

GO:CC  annotations: BioMart
classes: releases/2022-03-22

GO:BP  annotations: BioMart
classes: releases/2022-03-22

KEGG  KEGG FTP Release 2022-05-16

REAC  annotations: BioMart
classes: 2022-5-18

WP  20220510

TF  annotations: TRANSFAC Release 2021.3
classes: v2

MIRNA  Release 7.0

HPA  annotations: HPA website: 21-12-06
classes: script: 21-12-17

CORUM  03.09.2018 Corum 3.0

HP  annotations: hpo.annotations #12
classes: None

Explanation for bonferroni correction selection is here: https://www.biostars.org/p/267078/#:~:text=The%20Bonferroni%20correction%20is%20important,obtain%20an%20expected%20error%20rate.

#### Looking at upregulated and downregulated enrichment for cortex
prefiltered genes to be upregulated geenes or downregulated genes and then used same forloop as before
```{r}
# upregulated:
cell_types <- unique(het_cortex_DEGS$celltype) # total of 7
het_cortex_DEGS_UP <- het_cortex_DEGS %>% filter(diffexpressed == "UP")

cortex_UP_result_list <- list()

for (i in cell_types) {
  # filter genes by cell type ----------
  genes <- as.list(het_cortex_DEGS_UP %>% filter(celltype == i))
  # submit genes to pathway analysis function ----------
  fea_result_filt <- fea_DGE(genes = genes, organism = "mmusculus")
  # storing results ----------
  cortex_UP_result_list[[i]] <- fea_result_filt
  # set png to save plot to ----------
  png(paste0(here("results/seurat/Pathways_07/"), "FEA_pathways_up_", i, ".png"),
    width = 1000, height = 1000
  )
  # plot dotplot and add title based on input variable ----------
  plot <- bubbleplot(fea_result_filt) +
    ggtitle(paste0("Top 50 Upregulated Pathways Associated with Predicted Target Genes in ", i)) +
    theme(plot.title = element_text(hjust = 0.5, size = 12))
  # finish plotting ----------
  print(plot)
  dev.off()
}

save(cortex_UP_result_list, file = here("results/seurat/Pathways_07/cortex_het_fea_DGE_UP.Rdata"))

# downregulated:
cell_types <- unique(het_cortex_DEGS$celltype) # total of 7
het_cortex_DEGS_DOWN <- het_cortex_DEGS %>% filter(diffexpressed == "DOWN")
cortex_DOWN_result_list <- list()
for (i in cell_types) {
  # filter genes by cell type ----------
  genes <- as.list(het_cortex_DEGS_DOWN %>% filter(celltype == i))
  # submit genes to pathway analysis function ----------
  fea_result_filt <- fea_DGE(genes = genes, organism = "mmusculus")
  # storing results ----------
  cortex_DOWN_result_list[[i]] <- fea_result_filt
  # set png to save plot to ----------
  png(paste0(here("results/seurat/Pathways_07/"), "FEA_pathways_down_", i, ".png"),
    width = 1000, height = 1000
  )
  # plot dotplot and add title based on input variable ----------
  plot <- bubbleplot(fea_result_filt) +
    ggtitle(paste0("Top 50 Downregulated Pathways Associated with Predicted Target Genes in ", i)) +
    theme(plot.title = element_text(hjust = 0.5, size = 12))
  # finish plotting ----------
  print(plot)
  dev.off()
}

save(cortex_DOWN_result_list, file = here("results/seurat/Pathways_07/cortex_het_fea_DGE_DOWN.Rdata"))
```


#### Looking at upregulated and downregulated for kidney 
prefiltered genes to be upregulated geenes or downregulated genes and then used same forloop as before
```{r}
# upregulated:
cell_types <- unique(het_kidney_DEGS$celltype) # total of 14
het_kidney_DEGS_UP <- het_kidney_DEGS %>% filter(diffexpressed == "UP")
kidney_UP_result_list <- list()

for (i in cell_types) {
  # filter genes by cell type ----------
  genes <- as.list(het_kidney_DEGS_UP %>% filter(celltype == i))
  # submit genes to pathway analysis function ----------
  fea_result_filt <- fea_DGE(genes = genes, organism = "mmusculus")
  # storing results ----------
  kidney_UP_result_list[[i]] <- fea_result_filt
  # set png to save plot to ----------
  png(paste0(here("results/seurat/Pathways_07/"), "FEA_pathways_up_", i, ".png"),
    width = 1000, height = 1000
  )
  # plot dotplot and add title based on input variable ----------
  plot <- bubbleplot(fea_result_filt) +
    ggtitle(paste0("Top 50 Upregulated Pathways Associated with Predicted Target Genes in ", i)) +
    theme(plot.title = element_text(hjust = 0.5, size = 12))
  # finish plotting ----------
  print(plot)
  dev.off()
}

save(kidney_UP_result_list, file = here("results/seurat/Pathways_07/kidney_het_fea_DGE_UP.Rdata"))

# downregulated:
cell_types <- unique(het_kidney_DEGS$celltype) # total of 14
het_kidney_DEGS_DOWN <- het_kidney_DEGS %>% filter(diffexpressed == "DOWN")
kidney_DOWN_result_list <- list()

for (i in cell_types) {
  # filter genes by cell type ----------
  genes <- as.list(het_kidney_DEGS_DOWN %>% filter(celltype == i))
  # submit genes to pathway analysis function ----------
  fea_result_filt <- fea_DGE(genes = genes, organism = "mmusculus")
  # storing results ----------
  kidney_DOWN_result_list[[i]] <- fea_result_filt
  # set png to save plot to ----------
  png(paste0(here("results/seurat/Pathways_07/"), "FEA_pathways_down_", i, ".png"),
    width = 1000, height = 1000
  )
  # plot dotplot and add title based on input variable ----------
  plot <- bubbleplot(fea_result_filt) +
    ggtitle(paste0("Top 50 Downregulated Pathways Associated with Predicted Target Genes in ", i)) +
    theme(plot.title = element_text(hjust = 0.5, size = 12))
  # finish plotting ----------
  print(plot)
  dev.off()
}

save(kidney_DOWN_result_list, file = here("results/seurat/Pathways_07/kidney_het_fea_DGE_DOWN.Rdata"))
```




#### Running FEA for grofiler2 with Setbp1 target gene set:

```{r loading-setbp1-targets}
setbp1_genes <- read.csv(here("results/seurat/setbp1_targets.csv"))
setbp1_genes <- setbp1_genes[, -1]
setbp1_genes <- append(setbp1_genes, "Setbp1")
```


#### Looking at upregulated and downregulated for kidney
prefiltered genes to be upregulated genes or downregulated genes and then used same forloop as before
```{r}
# up and down side by side
cell_types <- unique(het_kidney_DEGS$celltype) # total of 14
for (i in cell_types) {
  # filter genes by cell type ----------
  genes <- het_kidney_DEGS %>% filter(celltype == i)
  # filter 'genes' input
  targets <- filter(genes, gene %in% setbp1_genes)

  # submit genes to pathway analysis function ----------
  if (nrow(targets) > 0) {
    fea_res <- combined_fea_set(genes = targets, organism = "mmusculus")

    # plot dotplot and add title based on input variable ----------
    png(paste0(here("results/seurat/Pathways_07/"), "FEA_SETBP1targetspathways_kidney_", i, ".png"),
      width = 1000, height = 3000
    )
    plot <- ggplot(fea_res$combined, aes(
      x = direction,
      y = reorder(term_name, -p_value),
      size = intersection_size,
      fill = p_value
    )) +
      geom_point(alpha = 0.7, shape = 21) +
      scale_size(range = c(2, 10), name = "Intersection Size") +
      scale_fill_distiller(palette = "Purples") +
      labs(x = "Direction", y = "Functional Enrichment Terms") +
      ggtitle(paste0("Pathways Associated with Differentially Expressed Target Genes in ", i)) +
      theme(plot.title = element_text(hjust = 0.5, size = 12))
    # finish plotting ----------
    print(plot)
    dev.off()
  } else {
    (print(paste("no DE targets in", i)))
  }
}
```


```{r}
# up and down side by side
cell_types <- unique(het_cortex_DEGS$celltype) # total of 14
for (i in cell_types) {
  # filter genes by cell type ----------
  genes <- het_cortex_DEGS %>% filter(celltype == i)
  # filter 'genes' input
  targets <- filter(genes, gene %in% setbp1_genes)

  # submit genes to pathway analysis function ----------
  if (nrow(targets) > 0) {
    fea_res <- combined_fea_set(genes = targets, organism = "mmusculus")
    # set png to save plot to ----------
    png(paste0(here("results/seurat/Pathways_07/"), "FEA_Setbp1targetspathways_cortex_", i, ".png"),
      width = 1000, height = 3000
    )
    # plot dotplot and add title based on input variable ----------
    plot <- ggplot(fea_res$combined, aes(
      x = direction,
      y = reorder(term_name, -p_value),
      size = intersection_size,
      fill = p_value
    )) +
      geom_point(alpha = 0.7, shape = 21) +
      scale_size(range = c(2, 10), name = "Intersection Size") +
      scale_fill_distiller(palette = "Purples") +
      labs(x = "Direction", y = "Functional Enrichment Terms") +
      ggtitle(paste0("Pathways Associated with Differentially Expressed Target Genes in ", i)) +
      theme(plot.title = element_text(hjust = 0.5, size = 12))
    # finish plotting ----------
    print(plot)
    dev.off()
  } else {
    (print(paste("no DE targets in", i)))
  } # none in mural_cells
}
```
Note: [1] "no DE targets in mural_cells"

```{r}
# run style
style_file(here("src/seurat_scripts/Setbp1_PathwayAnalysisCortexKidney_07.Rmd"))
# lintr was run as well
```

```{r}
proc.time()
```
    user   system  elapsed 
1243.685   40.958 2231.599


# Session Info
```{r}
sessionInfo()
```
R version 4.1.3 (2022-03-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] styler_1.9.0       lintr_3.0.2        here_1.0.1         rrvgo_1.6.0        msigdbr_7.5.1      lattice_0.20-45    lubridate_1.9.2    forcats_1.0.0      stringr_1.5.0     
[10] dplyr_1.1.0        purrr_1.0.1        readr_2.1.4        tidyr_1.3.0        tibble_3.1.8       ggplot2_3.4.1      tidyverse_2.0.0    gprofiler2_0.2.1   SeuratObject_4.1.3
[19] Seurat_4.3.0      

loaded via a namespace (and not attached):
  [1] plyr_1.8.8             igraph_1.4.1           lazyeval_0.2.2         sp_1.6-0               splines_4.1.3          listenv_0.9.0          scattermore_0.8       
  [8] gridBase_0.4-7         GenomeInfoDb_1.30.1    digest_0.6.31          GOSemSim_2.20.0        htmltools_0.5.4        GO.db_3.14.0           fansi_1.0.4           
 [15] magrittr_2.0.3         memoise_2.0.1          tm_0.7-11              tensor_1.5             cluster_2.1.2          ROCR_1.0-11            remotes_2.4.2         
 [22] tzdb_0.3.0             globals_0.16.2         Biostrings_2.62.0      wordcloud_2.6          matrixStats_0.63.0     R.utils_2.12.2         timechange_0.2.0      
 [29] spatstat.sparse_3.0-0  colorspace_2.1-0       treemap_2.4-3          blob_1.2.3             ggrepel_0.9.3          xfun_0.37              callr_3.7.3           
 [36] crayon_1.5.2           RCurl_1.98-1.10        jsonlite_1.8.4         progressr_0.13.0       spatstat.data_3.0-0    survival_3.3-1         zoo_1.8-11            
 [43] glue_1.6.2             polyclip_1.10-4        gtable_0.3.1           zlibbioc_1.40.0        XVector_0.34.0         leiden_0.4.3           R.cache_0.16.0        
 [50] future.apply_1.10.0    BiocGenerics_0.40.0    abind_1.4-5            scales_1.2.1           pheatmap_1.0.12        DBI_1.1.3              spatstat.random_3.1-3 
 [57] miniUI_0.1.1.1         Rcpp_1.0.10            viridisLite_0.4.1      xtable_1.8-4           reticulate_1.28        bit_4.0.5              stats4_4.1.3          
 [64] rex_1.2.1              htmlwidgets_1.6.1      httr_1.4.5             RColorBrewer_1.1-3     ellipsis_0.3.2         ica_1.0-3              R.methodsS3_1.8.2     
 [71] pkgconfig_2.0.3        uwot_0.1.14            deldir_1.0-6           utf8_1.2.3             tidyselect_1.2.0       rlang_1.0.6            reshape2_1.4.4        
 [78] later_1.3.0            AnnotationDbi_1.56.2   munsell_0.5.0          tools_4.1.3            cachem_1.0.7           cli_3.6.0              generics_0.1.3        
 [85] RSQLite_2.3.0          ggridges_0.5.4         evaluate_0.20          fastmap_1.1.1          yaml_2.3.7             goftest_1.2-3          processx_3.8.0        
 [92] babelgene_22.9         knitr_1.42             bit64_4.0.5            fitdistrplus_1.1-8     RANN_2.6.1             KEGGREST_1.34.0        pbapply_1.7-0         
 [99] future_1.31.0          nlme_3.1-155           mime_0.12              slam_0.1-50            R.oo_1.25.0            xml2_1.3.3             compiler_4.1.3        
[106] rstudioapi_0.14        plotly_4.10.1          png_0.1-8              spatstat.utils_3.0-1   stringi_1.7.12         ps_1.7.2               cyclocomp_1.1.0       
[113] desc_1.4.2             Matrix_1.5-3           vctrs_0.5.2            pillar_1.8.1           lifecycle_1.0.3        spatstat.geom_3.0-6    lmtest_0.9-40         
[120] RcppAnnoy_0.0.20       data.table_1.14.8      cowplot_1.1.1          bitops_1.0-7           irlba_2.3.5.1          httpuv_1.6.9           patchwork_1.1.2       
[127] R6_2.5.1               promises_1.2.0.1       KernSmooth_2.23-20     gridExtra_2.3          IRanges_2.28.0         parallelly_1.34.0      codetools_0.2-18      
[134] MASS_7.3-55            rprojroot_2.0.3        withr_2.5.0            sctransform_0.3.5      S4Vectors_0.32.4       GenomeInfoDbData_1.2.7 parallel_4.1.3        
[141] hms_1.1.2              grid_4.1.3             rmarkdown_2.20         Rtsne_0.16             spatstat.explore_3.0-6 NLP_0.2-1              Biobase_2.54.0        
[148] shiny_1.7.4           

This is the final script for all Seurat preprocessing and expression investigation. The next set of scripts are located in 'src/networks/'
