---
title: "TF_motif_reconstruction"
output: html_document
date: '2023-01-23'
output:     
  html_document:
      toc: true
      toc_depth: 2
      toc_float: true
---

Goal: Reconstruct TF-motif input to include additional resources and be more comprehensive. 
*See detailed overview here: https://docs.google.com/document/d/1se-hztBXdYIbF0T0E9zNQFXX6Cib3F_KvdzrOId0NI4/edit?usp=sharing

# Setup
```{r loading-libraries, results='hide'}
library(readr)
library(dplyr)
library(here)
library(styler)
library(lintr)
set.seed(2178)
```

# Create Human (Homo Sapiens) Input:
Note: The previous TF_motif input was generated in the 'Setbp1_PANDA_TFmotif.Rmd'
```{r load-previous-TF-motif}
hu_motif <- read.table(file = here("data/processed/motif_inputs/hg38_TFmotifs.txt"), sep = "\t") # load in motif data
```

```{r load-human-SIGNOR}
signor_hu <- read_tsv(file = here("data/unprocessed/SIGNOR_all_human_data_23_01_23.tsv"), col_names = TRUE) # load in motif data

# filter for only "trasncriptional regulation" and score >= 0.4, and proteins:
signor_hu <- signor_hu[signor_hu$MECHANISM == "transcriptional regulation", ] %>%
  filter(., rowSums(is.na(.)) != ncol(.)) %>%
  filter(., TYPEA == "protein") %>%
  filter(., TYPEB == "protein") %>%
  filter(., SCORE >= 0.4, )

# grab ENTITYA, ENTITYB, and SCORE columns (correspond to TF, Motif, and Score)
signor_hu <- signor_hu[, c(1, 5, 27)]
colnames(signor_hu) <- c("TF_Info", "Motif", "interaction")
signor_hu$interaction <- 1
```

```{r load-human-TRRUST}
trrust_hu <- read_tsv(file = here("data/unprocessed/trrust_rawdata.human.tsv"), col_names = "X") # load in motif data

colnames(trrust_hu) <- c("TF_Info", "Motif", "mechanism", "PMID")
trrust_hu <- trrust_hu[, c(1:2)]
trrust_hu$interaction <- 1
```

```{r merge-all}
hu_motif_all <- unique(rbind(hu_motif, signor_hu) %>% rbind(., trrust_hu))
```

# Create Mouse (Musc Musculus) Input:
Note: The previous TF_motif input was generated in the 'Setbp1_PANDA_TFmotif.Rmd'
```{r load-previous-TF-motif}
mus_motif <- read.table(file = here("data/processed/motif_inputs/mm10_TFmotifs.txt"), sep = "\t") # load in motif data
```

```{r load-mouse-SIGNOR}
signor_mus <- read_tsv(file = here("data/unprocessed/SIGNOR_all_data_M_musculus_23_01_23.tsv"), col_names = TRUE) # load in motif data

# filter for only "trasncriptional regulation" and score >= 0.4, and proteins:
signor_mus <- signor_mus[signor_mus$MECHANISM == "transcriptional regulation", ] %>%
  filter(., rowSums(is.na(.)) != ncol(.)) %>%
  filter(., TYPEA == "protein") %>%
  filter(., TYPEB == "protein") %>%
  filter(., SCORE >= 0.4, )

# grab ENTITYA, ENTITYB, and SCORE columns (correspond to TF, Motif, and Score)
signor_mus <- signor_mus[, c(1, 5, 27)]
colnames(signor_mus) <- c("TF_Info", "Motif", "interaction")
signor_mus$interaction <- 1
```

```{r load-mouse-TRRUST}
trrust_mus <- read_tsv(file = here("data/unprocessed/trrust_rawdata.mouse.tsv"), col_names = "X") # load in motif data

colnames(trrust_mus) <- c("TF_Info", "Motif", "mechanism", "PMID")
trrust_mus <- trrust_mus[, c(1:2)]
trrust_mus$interaction <- 1
```

```{r merge-all}
mus_motif_all <- unique(rbind(mus_motif, signor_mus) %>% rbind(., trrust_mus))
```

```{r manually-add-SETBP1-chipseq}
# load in Setbp1 data from MSIGDB:
SETBP1geneset_syngo <- read_csv(here("data/processed/msigdb/SETBP1genset_syngo.csv"))
msig <- matrix(NA, nrow = 188, ncol = 3)
msig <- data.frame(msig)
colnames(msig) <- c("TF_Info", "Motif", "interaction")
msig$Motif <- SETBP1geneset_syngo$query # add in geneset data
msig$TF_Info <- "SETBP1"
msig$interaction <- 1

# merge
hu_motif_all <- unique(rbind(hu_motif_all, msig))
```

```{r manually-add-Setbp1-chipseq}
# load in Setbp1 data from MSIGDB:
SETBP1geneset_syngo <- read_csv(here("data/processed/msigdb/SETBP1genset_syngo.csv"))
mus_msig <- matrix(NA, nrow = 157, ncol = 3)
mus_msig <- data.frame(mus_msig)
colnames(mus_msig) <- c("TF_Info", "Motif", "interaction")
genes <- SETBP1geneset_syngo[!(is.na(SETBP1geneset_syngo$symbol)), 4]
mus_msig$Motif <- genes$symbol # add in geneset data
mus_msig$TF_Info <- "Setbp1"
mus_msig$interaction <- 1

# merge
mus_motif_all <- unique(rbind(mus_motif_all, mus_msig))
```

# Save Outputs once:
```{r save-human}
write.table(hu_motif_all, file = here("data/processed/motif_inputs/hu_motif_all.txt"), sep = "\t")
```

```{r save-mouse}
write.table(mus_motif_all, file = here("data/processed/motif_inputs/mus_motif_all.txt"), sep = "\t")
```

```{r}
# run style
style_file(here("src/network_scripts/input_construction/HumanMouse_TF_motif_reconstructed.Rmd"))
# lintr was run as well
```

```{r}
proc.time()
```
   user  system elapsed 
143.345  25.066 449.363

```{r}
sessionInfo()
```
R version 4.1.3 (2022-03-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] lintr_3.0.2  styler_1.9.0 here_1.0.1   dplyr_1.1.0  readr_2.1.4 

loaded via a namespace (and not attached):
 [1] pillar_1.8.1      compiler_4.1.3    R.methodsS3_1.8.2 R.utils_2.12.2    remotes_2.4.2     tools_4.1.3       bit_4.0.5         digest_0.6.31    
 [9] evaluate_0.20     lifecycle_1.0.3   tibble_3.1.8      R.cache_0.16.0    pkgconfig_2.0.3   rlang_1.0.6       rex_1.2.1         cli_3.6.0        
[17] rstudioapi_0.14   yaml_2.3.7        parallel_4.1.3    xfun_0.37         fastmap_1.1.1     cyclocomp_1.1.0   xml2_1.3.3        withr_2.5.0      
[25] knitr_1.42        generics_0.1.3    desc_1.4.2        vctrs_0.5.2       hms_1.1.2         bit64_4.0.5       rprojroot_2.0.3   tidyselect_1.2.0 
[33] glue_1.6.2        R6_2.5.1          processx_3.8.0    fansi_1.0.4       vroom_1.6.1       rmarkdown_2.20    tzdb_0.3.0        purrr_1.0.1      
[41] callr_3.7.3       magrittr_2.0.3    htmltools_0.5.4   ps_1.7.2          ellipsis_0.3.2    utf8_1.2.3        lazyeval_0.2.2    crayon_1.5.2     
[49] R.oo_1.25.0  

